File: C/141. Linked List Cycle.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
bool hasCycle(struct ListNode *head) {
    bool traverse(struct ListNode *fast, struct ListNode *slow){
        if(fast == NULL || fast->next == NULL){
            return false;
        }

        if(slow == NULL || slow->next == NULL){
            return false;
        }

        if(slow == fast){
            return true;
        }

        return traverse(fast->next->next, slow->next);
    }
    

    if(head == NULL){
        return false;
    }

    return traverse(head->next, head);
}

File: Java/141. Linked List Cycle.java
import java.util.HashMap;
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public boolean hasCycle(ListNode head) {
        
        HashSet<ListNode> dict = new HashSet<>();

        while(head != null){
            if(dict.contains(head)){
                return true;
            } else {
                dict.add(head);
                head = head.next;
            }
            
        }
        return false;
    }
}
